File: backend/src/main/java/com/back/ourlog/global/security/oauth/OAuthAttributes.java
package com.back.ourlog.global.security.oauth;

import lombok.Getter;

import java.util.Map;

@Getter
public class OAuthAttributes {

    private Map<String, Object> attributes;
    private String nameAttributeKey;
    private String providerId;
    private String name;
    private String email;
    private String provider;

    public static OAuthAttributes of(String registrationId, String userNameAttributeName, Map<String, Object> attributes) {
        if ("naver".equals(registrationId)) {
            return ofNaver("id", (Map<String, Object>) attributes.get("response"));
        } else if ("kakao".equals(registrationId)) {
            return ofKakao("id", attributes);
        }

        return ofGoogle(userNameAttributeName, attributes);
    }

    private static OAuthAttributes ofGoogle(String userNameAttributeName, Map<String, Object> attributes) {
        return new OAuthAttributes(
                attributes,
                userNameAttributeName,
                (String) attributes.get("sub"),  // 고유 ID 필드: "sub"
                (String) attributes.get("name"),
                (String) attributes.get("email"),
                "google"
        );
    }

    private static OAuthAttributes ofNaver(String userNameAttributeName, Map<String, Object> attributes) {
        return new OAuthAttributes(
                attributes,
                userNameAttributeName,
                (String) attributes.get("id"),
                (String) attributes.get("name"),
                (String) attributes.get("email"),
                "naver"
        );
    }

    private static OAuthAttributes ofKakao(String userNameAttributeName, Map<String, Object> attributes) {
        Map<String, Object> kakaoAccount = (Map<String, Object>) attributes.get("kakao_account");
        Map<String, Object> profile = (Map<String, Object>) kakaoAccount.get("profile");

        return new OAuthAttributes(
                attributes,
                userNameAttributeName,
                String.valueOf(attributes.get("id")),
                (String) profile.get("nickname"),
                (String) kakaoAccount.get("email"),
                "kakao"
        );
    }

    public OAuthAttributes(Map<String, Object> attributes, String nameAttributeKey, String providerId, String name, String email, String provider) {
        this.attributes = attributes;
        this.nameAttributeKey = nameAttributeKey;
        this.providerId = providerId;
        this.name = name;
        this.email = email;
        this.provider = provider;
    }
}

File: backend/src/main/java/com/back/ourlog/global/security/oauth/OAuthAttributes.kt
package com.back.ourlog.global.security.oauth

data class OAuthAttributes(
    val attributes: Map<String, Any>,
    val providerId: String,
    val name: String,
    val email: String,
    val provider: String
) {
    companion object {

        fun of(registrationId: String, attributes: Map<String, Any>): OAuthAttributes {
            return when (registrationId.lowercase()) {
                "naver" -> ofNaver(attributes["response"] as? Map<*, *>)
                "kakao" -> ofKakao(attributes)
                else -> ofGoogle(attributes)
            }
        }

        private fun ofGoogle(attributes: Map<String, Any>): OAuthAttributes {
            val providerId = attributes["sub"]?.toString().orEmpty()
            val name = attributes["name"]?.toString().orEmpty()
            val email = attributes["email"]?.toString().orEmpty()
            val profileImageUrl = attributes["picture"]?.toString() // Google's profile picture URL

            val modifiedAttributes = attributes.toMutableMap()
            if (profileImageUrl != null) {
                modifiedAttributes["profileImageUrl"] = profileImageUrl
            }

            return OAuthAttributes(
                attributes = modifiedAttributes as Map<String, Any>,
                providerId = providerId,
                name = name,
                email = email,
                provider = "google"
            )
        }

        private fun ofNaver(response: Map<*, *>?): OAuthAttributes {
            val resp = response ?: emptyMap<Any, Any>()
            val providerId = resp["id"]?.toString().orEmpty()
            val name = resp["name"]?.toString().orEmpty()
            val email = resp["email"]?.toString().orEmpty()
            val profileImageUrl = resp["profile_image"]?.toString() // Naver's profile image URL

            val modifiedAttributes = resp.mapKeys { it.key.toString() }.toMutableMap()
            if (profileImageUrl != null) {
                modifiedAttributes["profileImageUrl"] = profileImageUrl
            }

            return OAuthAttributes(
                attributes = modifiedAttributes as Map<String, Any>,
                providerId = providerId,
                name = name,
                email = email,
                provider = "naver"
            )
        }

        private fun ofKakao(attributes: Map<String, Any>): OAuthAttributes {
            val kakaoAccount = attributes["kakao_account"] as? Map<*, *> ?: emptyMap<Any, Any>()
            val profile = kakaoAccount["profile"] as? Map<*, *> ?: emptyMap<Any, Any>()

            val providerId = attributes["id"]?.toString().orEmpty()
            val name = profile["nickname"]?.toString().orEmpty()
            val email = kakaoAccount["email"]?.toString() ?: "$providerId@kakao.com"
            val profileImageUrl = profile["profile_image_url"]?.toString() // Kakao's profile image URL

            val modifiedAttributes = attributes.toMutableMap()
            if (profileImageUrl != null) {
                modifiedAttributes["profileImageUrl"] = profileImageUrl
            }

            return OAuthAttributes(
                attributes = modifiedAttributes as Map<String, Any>,
                providerId = providerId,
                name = name,
                email = email,
                provider = "kakao"
            )
        }
    }
}